### ISSUE TITLE
[bug]: Seamless Tile not  seamless on certain colors

### ISSUE THREAD

**Issue by KingsmanZer0** [invoke-ai/InvokeAI#6340]
### Is there an existing issue for this problem?

- [X] I have searched the existing issues

### Operating system

Windows

### GPU vendor

Nvidia (CUDA)

### GPU model

rtx 4090

### GPU VRAM

24

### Version number

4.2

### Browser

chrome 124.0.6367

### What happened

trying to generate seamless tiles but when generating images with light red/pink/blue colors there is a visible light brown line going down the seams after i try the tile in photoshop. tried generating a couple that has light reds/blues and issue still persist.
also tried a couple of other colors like white or dark colors but seams were fine and seamless
![seams1](https://github.com/invoke-ai/InvokeAI/assets/93020987/20e069e0-a374-4520-baca-0cd708fb3b4d)

### What you expected to happen

should be seamless transitions in colors

### How to reproduce the problem

Prompt used: red flower watercolor
Model: juggernautxl9
size: 1024 x 1024
Enabled:  X and Y seamless tile

### Additional context

When trying to generate seamless tile an images with light red/pink/blue colors a visible light brown seam is visible. tried some other colors like white or dark colors but they seemed fine.
![seams2](https://github.com/invoke-ai/InvokeAI/assets/93020987/fb4defd4-5ea5-4f0e-bca5-6b5cd25f08fc)

### Discord username

tuanny87

**Comment by psychedelicious** [comment]
I can't seem to reproduce this, or maybe I just don't see the issue. I've made dozens of images with x & y seamless, the same model and same or similar prompts. Can't see the line on any of them.

Here's a selection where the petals cross the central vertical axis of the image, where the line is visible:

<table width="100%">
  <tbody>
    <tr>
      <td width="20%"><img src="https://github.com/invoke-ai/InvokeAI/assets/4822129/615b8aa8-ac40-44a4-8c39-38193a2075a0"/></td>
      <td width="20%"><img src="https://github.com/invoke-ai/InvokeAI/assets/4822129/af25ce78-41e3-4c1b-8a37-40059e4098c2"/></td>
      <td width="20%"><img src="https://github.com/invoke-ai/InvokeAI/assets/4822129/5c38ef93-1750-40a4-aa51-7dfa52781672"/></td>
      <td width="20%"><img src="https://github.com/invoke-ai/InvokeAI/assets/4822129/7bec1ba2-0333-4ca4-af34-1f31d3005241"/></td>
      <td width="20%"><img src="https://github.com/invoke-ai/InvokeAI/assets/4822129/66dd44d5-342b-46cd-91c8-dd1ccf96ded7"/></td>
    </tr>
  </tbody>
</table>

Do you see any lines on these?

**Comment by KingsmanZer0** [comment]
> I can't seem to reproduce this, or maybe I just don't see the issue. I've made dozens of images with x & y seamless, the same model and same or similar prompts. Can't see the line on any of them.
> 
> Here's a selection where the petals cross the central vertical axis of the image, where the line is visible:
> 
> ![](https://private-user-images.githubusercontent.com/4822129/329520784-615b8aa8-ac40-44a4-8c39-38193a2075a0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUzMzI3MTgsIm5iZiI6MTcxNTMzMjQxOCwicGF0aCI6Ii80ODIyMTI5LzMyOTUyMDc4NC02MTViOGFhOC1hYzQwLTQ0YTQtOGMzOS0zODE5M2EyMDc1YTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MTBUMDkxMzM4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjA3NTBjMmJhYTNlN2UwMWM5Y2I2ZWFhNWE4ZjY3MDVjMTU1N2UxYjAzYTZhMDQzYWRiOTZkYTA4NDk1Y2MxZSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.P55tkReGh1uJZfPmb831iJyeOoocefRsRhjSdfcqs2I)	![](https://private-user-images.githubusercontent.com/4822129/329520906-af25ce78-41e3-4c1b-8a37-40059e4098c2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUzMzI3MTgsIm5iZiI6MTcxNTMzMjQxOCwicGF0aCI6Ii80ODIyMTI5LzMyOTUyMDkwNi1hZjI1Y2U3OC00MWUzLTRjMWItOGEzNy00MDA1OWU0MDk4YzIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MTBUMDkxMzM4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2Y0ZWU4ZTkyMzIzNmNjZDgyYmY5OWQ5YzM3NTJkOGNhMzFiMjc4ODZiYWU0NzFmMGVmY2VjNzE1YTllOGIzMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.UjA_Og5rPR-KiTD1gfVaXAszOlVyPik2q9sy_y1GXrA)	![](https://private-user-images.githubusercontent.com/4822129/329520929-5c38ef93-1750-40a4-aa51-7dfa52781672.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUzMzI3MTgsIm5iZiI6MTcxNTMzMjQxOCwicGF0aCI6Ii80ODIyMTI5LzMyOTUyMDkyOS01YzM4ZWY5My0xNzUwLTQwYTQtYWE1MS03ZGZhNTI3ODE2NzIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MTBUMDkxMzM4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzM4YmE2NGU3YmU2ZmRiMzM1ZjQ3NWI2OGIzY2VkYjQwZWQ1ZDFjOTk1ZjQ2NTI0NGRkY2YzOWNhY2NmOGJlOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.1vPkM1YuwUgWcMbxowlGvkTU_Yd8YPzfqELopZaUT-4)	![](https://private-user-images.githubusercontent.com/4822129/329520950-7bec1ba2-0333-4ca4-af34-1f31d3005241.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUzMzI3MTgsIm5iZiI6MTcxNTMzMjQxOCwicGF0aCI6Ii80ODIyMTI5LzMyOTUyMDk1MC03YmVjMWJhMi0wMzMzLTRjYTQtYWYzNC0xZjMxZDMwMDUyNDEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MTBUMDkxMzM4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTQzYjFkYWI4MWVlZDllMTdhNDZiNTkxNTMwYzU2NTI5M2I3MzYwMzMzNDgyOWVjZmQ2MmI2ZGUzN2EzZWRhOCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.2rncxNcWmSiHIjtoLNPyxiZOn7D-Sah3yr0U7DMnmyQ)	![](https://private-user-images.githubusercontent.com/4822129/329520981-66dd44d5-342b-46cd-91c8-dd1ccf96ded7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTUzMzI3MTgsIm5iZiI6MTcxNTMzMjQxOCwicGF0aCI6Ii80ODIyMTI5LzMyOTUyMDk4MS02NmRkNDRkNS0zNDJiLTQ2Y2QtOTFjOC1kZDFjY2Y5NmRlZDcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUxMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MTBUMDkxMzM4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWNlMmQyZWE3ODIwMGM3YjEyMmVlOWExZmUyODg1ODA5NjUxZTEwZTRkMjA2NDk2ODE1YWMzMjcwMzc2ZGVhNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.1Cg2mVAn_kWq1gwacR39-4yvVNQemrbQRmyYMoXY2tM)
> Do you see any lines on these?

you have to join the tile in photoshop to see the seam issue it. here i took your first image you generated and made a duplicate and joined you can see the brown line on the flower
![seam line](https://github.com/invoke-ai/InvokeAI/assets/93020987/f4d6eec8-dcc0-4709-86f0-2625f91ea5cc)

**Comment by psychedelicious** [comment]
Ahh, sorry, I misunderstood the issue.

@RyanJDick @blessedcoolant I wonder if our seamless isn't quite right yet. I noticed that A1111 appears to do [fully-circular tiling only](https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/1c0a0c4c26f78c32095ebc7f8af82f5c04fca8c0/modules/sd_hijack.py#L366). It's done by patching `torch.nn.Conv2d`'s constructor. Very simple, but doesn't support tiling on individual axes.

There's a long-running diffusers issue about seamless, and here's the implementation that apparently was settled-on as working: https://github.com/huggingface/diffusers/issues/556#issuecomment-1993287019

This is similar to [ours](https://github.com/invoke-ai/InvokeAI/blob/main/invokeai/backend/stable_diffusion/seamless.py), with some notable changes:
- We skip some layers
- We also patch `nn.ConvTranspose2d`
- The diffusers implementation has special handling for `LoRACompatibleConv` layers

I don't understand this well enough to making any changes that aren't just guesses/experimentation.

**Comment by blessedcoolant** [comment]
@psychedelicious https://github.com/invoke-ai/InvokeAI/pull/6344 Implemented here. Seems to be working. Give it a run.

**Comment by psychedelicious** [comment]
@blessedcoolant I'm asking why we skip a layer and handle the transpose layer, because the comments in the code seem to indicate that is important. At some point in the past, our implementation was like the diffusers one and we changed it to what we have now.

**Comment by blessedcoolant** [comment]
> @blessedcoolant I'm asking why we skip a layer and handle the transpose layer, because the comments in the code seem to indicate that is important. At some point in the past, our implementation was like the diffusers one and we changed it to what we have now.

Because technically we also need to change the layers in the text_encoder for seamless .. but we just do UNet and VAE. So I'm guessing skipping the last layer was a sort of "fix" to mitigate some of the prompt importance in favor of seamlessness. I cannot really tell why it was done that way.

So far in my testing, I haven't found any cases where this new/old algo has failed. It's a lot simpler and clearer on what it does.

And I don't know why there was even a check for the transposed layers coz there's none in the UNet or the VAE.

**Comment by psychedelicious** [comment]
Ok. What if we patch torch directly like A1111 does? That would automatically apply to every model.

**Comment by blessedcoolant** [comment]
@psychedelicious He doesn't actually use that function. He just applies circular padding to the layers as per requirement of the generation.  https://github.com/AUTOMATIC1111/stable-diffusion-webui/blob/1c0a0c4c26f78c32095ebc7f8af82f5c04fca8c0/modules/sd_hijack.py#L311

**Comment by psychedelicious** [comment]
@blessedcoolant Ahhh I see. Thanks for clearing that up.
